File: webapp/server-node/src/authz.ts
/**
 * authz.ts — CROSS-CUTTING: single permission choke point (G0.1, ADR-006)
 * ===========================================================================
 * EVERY authorization decision goes through `can()`. The matrix is data, so
 * it is exhaustively unit-testable (test/p0.test.ts walks every role ×
 * action pair). Route guards are thin wrappers around this module.
 *
 * MENTAL MODEL FOR NEW TEAM MEMBERS
 * ---------------------------------
 * 1. A request arrives with a Bearer token → `authenticate()` (auth.ts)
 *    resolves WHO is calling (userId).
 * 2. `buildActorContext()` gathers WHAT that person IS: every ACTIVE persona
 *    row (G0.1b: a person can be moderator + crowd_expert + learner at once)
 *    plus WHICH farms they own or are a member of.
 * 3. `can(ctx, action, resource)` answers the single question "may this
 *    actor perform this action on this resource?" — nothing else in the
 *    codebase decides permissions (single choke point rule).
 * 4. `requirePermission(action)` wraps steps 2–3 as a Fastify preHandler and
 *    attaches `request.actor` for handlers that need farm scoping.
 *
 * FAIL-SAFE RULES (security posture)
 * ----------------------------------
 * - Unknown action names are DENIED even for admins (fail closed).
 * - Missing/unknown farmId on farm-scoped actions is DENIED.
 * - Every denial is logged (debug in dev; warn for security-relevant ones)
 *   so production dashboards can alert on abnormal denial rates.
 *
 * REQUIREMENT TRACEABILITY
 * ------------------------
 *   - docs/V2_REQUIREMENTS_ANALYSIS.md §G0.1 (roles & visibility matrix)
 *   - docs/V2_REQUIREMENTS_ANALYSIS.md §G0.1b (multi-persona identity model)
 *   - docs/READINESS_REVIEW.md §3 guarantee #1 (server-side enforcement;
 *     client hiding is cosmetic only)
 */

import type { FastifyReply, FastifyRequest } from 'fastify';
import type { IssueStage, Persona } from './types.js';
import { getUser, listUserPersonas, listFarmMembers } from './store.js';
import { makeLogger } from './logger.js';

// Scoped logger: every authorization decision can be traced in production
// with `grep '"scope":"authz"'`. Debug-level = normal flow; warn = denials.
const log = makeLogger('authz');

/** Everything an authorization check may need about the caller. */
export interface ActorContext {
  userId: string;
  /** Union of all active personas (G0.1b multi-persona model). */
  personas: Persona[];
  /** Farms this user OWNS. */
  ownedFarmIds: Set<string>;
  /** farmId → roleInFarm for memberships. */
  memberships: Map<string, 'owner' | 'moderator' | 'worker' | 'accountant'>;
}

/** Actions protected by the matrix. Extend as modules land. */
export type Action =
  | 'issue.view'
  | 'issue.create'
  | 'issue.advance'
  | 'issue.close'
  | 'task.create'
  /** Any-authenticated reads whose farm scoping happens IN THE HANDLER via
   *  assertFarmAccess() (generic lists like devices/videos/trees). */
  | 'device.view'
  | 'valve.control'
  | 'plan.manage'
  | 'subscription.assign'
  | 'persona.verify'
  | 'audit.view'
  | 'flag.manage';

/** Resources a guard may scope the check to. */
export interface ResourceRef {
  farmId?: string;
  issueStage?: IssueStage;
}

/** Resolve the full ActorContext for a user (store lookups + persona union). */
export function buildActorContext(userId: string): ActorContext | null {
  const user = getUser(userId);
  if (!user) return null;
  // Primary role is seeded as a persona row too; read both sources and union.
  const held = new Set<Persona>([user.role as Persona]);
  for (const p of listUserPersonas(userId)) {
    if (p.status === 'active') held.add(p.persona);
  }
  const owned = new Set<string>();
  const memberships = new Map<string, 'owner' | 'moderator' | 'worker' | 'accountant'>();
  for (const m of listFarmMembers(userId)) {
    if (m.roleInFarm === 'owner') owned.add(m.farmId);
    else memberships.set(m.farmId, m.roleInFarm);
  }
  return { userId, personas: [...held], ownedFarmIds: owned, memberships };
}

/** Every action `can()` knows about — unknown ones must FAIL CLOSED, even for admins. */
const KNOWN_ACTIONS: Action[] = [
  'issue.view', 'issue.create', 'issue.advance', 'issue.close', 'task.create',
  'device.view', 'valve.control',
  'plan.manage', 'subscription.assign', 'persona.verify', 'audit.view', 'flag.manage',
];

/** Is `ctx` allowed `action` on `resource`? THE decision function. */
export function can(ctx: ActorContext, action: Action, resource: ResourceRef = {}): boolean {
  if (!(KNOWN_ACTIONS as string[]).includes(action)) return false; // unknown → deny
  if (ctx.personas.includes('admin')) return true;

  switch (action) {
    // Farm-scoped visibility: any member (or the owner) of THAT farm.
    case 'issue.view':
      return belongsToFarm(ctx, resource.farmId);

    // Authenticated read — real scoping is enforced per-item in handlers
    // through assertFarmAccess(); this keeps generic lists usable while
    // individual resources stay protected.
    case 'device.view':
      return ctx.personas.length > 0;

    // Workers may REPORT problems; moderators/owners manage their farm's issues.
    case 'issue.create':
      return (
        belongsToFarmWithRole(ctx, resource.farmId, ['worker', 'moderator']) ||
        ctx.ownedFarmIds.has(resource.farmId ?? '')
      );

    // Advancing stages is further constrained per-stage inside modules/issues.ts
    case 'issue.advance':
      return (
        belongsToFarmWithRole(ctx, resource.farmId, ['worker', 'moderator']) ||
        ctx.ownedFarmIds.has(resource.farmId ?? '')
      );

    // Physical actuation is high-stakes: moderator+ ONLY (never workers).
    // Full audit trail is written by the route (ADR-013-adjacent safety rule).
    case 'valve.control':
      return (
        belongsToFarmWithRole(ctx, resource.farmId, ['moderator']) ||
        ctx.ownedFarmIds.has(resource.farmId ?? '')
      );

    // Closing requires authority over the farm.
    case 'issue.close':
      return (
        belongsToFarmWithRole(ctx, resource.farmId, ['moderator']) ||
        ctx.ownedFarmIds.has(resource.farmId ?? '')
      );

    // Platform administration is admin-only by definition.
    case 'plan.manage':
    case 'subscription.assign':
    case 'persona.verify':
    case 'audit.view':
    case 'flag.manage':
      return false;

    default:
      return false; // deny unknown actions — fail closed
  }
}

// --- helpers ----------------------------------------------------------------

/**
 * Handler-side tenancy assertion for item-level reads: the caller must own,
 * be a member of, or administer THAT farm. Throws nothing — returns boolean;
 * routes map false → 403.
 */
export function hasFarmAccess(ctx: ActorContext, farmId?: string): boolean {
  return !!farmId && belongsToFarm(ctx, farmId);
}

function belongsToFarm(ctx: ActorContext, farmId?: string): boolean {
  if (!farmId) return false;
  return ctx.ownedFarmIds.has(farmId) || ctx.memberships.has(farmId);
}

function belongsToFarmWithRole(
  ctx: ActorContext,
  farmId: string | undefined,
  roles: Array<'worker' | 'moderator' | 'accountant' | 'owner'>
): boolean {
  if (!farmId) return false;
  const r = ctx.memberships.get(farmId);
  return r !== undefined && roles.includes(r);
}

/**
 * Fastify preHandler factory used by /v2 routes:
 * resolves the session → ActorContext, evaluates `can()` (skipped when no
 * action given = auth-only), exposes both as request.session / request.actor.
 * Sends 401/403 itself otherwise.
 */
export function requirePermission(action?: Action, getResource?: (request: FastifyRequest) => ResourceRef) {
  return async (request: FastifyRequest, reply: FastifyReply) => {
    const session = authenticateSession(request);
    if (!session) {
      await reply.code(401).send({ error: 'Unauthorized' });
      return;
    }
    const actor = buildActorContext(session.userId);
    if (!actor) {
      await reply.code(401).send({ error: 'Unknown user' });
      return;
    }
    if (action) {
      const resource = getResource ? getResource(request) : {};
      if (!can(actor, action, resource)) {
        // Denials are security signals: warn-level so they survive 'info'
        // production configs and can drive brute-force/IDOR alerts.
        log.warn('permission denied', {
          action,
          userId: session.userId,
          personas: actor.personas,
          resource,
        });
        await reply.code(403).send({ error: `Forbidden: ${action}` });
        return;
      }
      // Debug trace of every GRANTED decision — only visible with LOG_LEVEL=debug.
      log.debug('permission granted', { action, userId: session.userId, resource });
    }
    (request as any).session = session;
    (request as any).actor = actor;
  };
}

/** Inline token verification shared with auth.ts (kept DRY via import). */
import { authenticate } from './auth.js';
function authenticateSession(request: FastifyRequest) {
  return authenticate(request);
}

File: webapp/server-rust/src/authz.rs
//! authz.rs — single permission choke point (G0.1 / G0.1b / ADR-017)
//! ===========================================================================
//! EXACT mirror of `server-node/src/authz.ts` semantics:
//!   • permissions = UNION of all ACTIVE personas (multi-persona model)
//!   • farm scoping via ownership + memberships
//!   • admin bypass — BUT unknown actions still FAIL CLOSED
//!   • `valve.control` is moderator+ ONLY (physical actuation rule, ADR-017)
//!
//! REQUIREMENT TRACEABILITY: V2_REQUIREMENTS_ANALYSIS §G0.1/§G0.1b,
//! READINESS_REVIEW §3 guarantee #1, IMPLEMENTATION_PLAN test matrix.

use crate::store::Db;
use crate::types::Persona;

/// Everything the decision function needs about one caller.
pub struct ActorContext {
    pub user_id: String,
    pub personas: Vec<Persona>,
    pub owned_farm_ids: Vec<String>,
    /// (farmId, roleInFarm) pairs for non-owner memberships.
    pub memberships: Vec<(String, String)>,
}

/// Actions protected by the matrix. Keep in lockstep with authz.ts!
#[derive(Clone, Copy, PartialEq, Debug)]
pub enum Action {
    IssueView,
    IssueCreate,
    IssueAdvance,
    IssueClose,
    DeviceView,
    ValveControl,
    PlanManage,
    SubscriptionAssign,
    PersonaVerify,
    AuditView,
    FlagManage,
}

/// Resolve the full context for a user (personas ∪ farms).
pub fn build_actor_context(db: &Db, user_id: &str) -> Option<ActorContext> {
    if !db.users.contains_key(user_id) {
        return None;
    }
    let mut personas: Vec<Persona> = Vec::new();
    // Primary role counts as a persona row too.
    if let Some(u) = db.users.get(user_id) {
        personas.push(u.role.clone());
    }
    for p in &db.personas {
        if p.user_id == user_id && p.status == "active" && !personas.contains(&p.persona) {
            personas.push(p.persona.clone());
        }
    }
    let mut owned = Vec::new();
    let mut memberships = Vec::new();
    for m in &db.farm_members {
        if m.user_id == user_id {
            if m.role_in_farm == "owner" {
                owned.push(m.farm_id.clone());
            } else {
                memberships.push((m.farm_id.clone(), m.role_in_farm.clone()));
            }
        }
    }
    Some(ActorContext { user_id: user_id.into(), personas, owned_farm_ids: owned, memberships })
}

fn belongs_to_farm(ctx: &ActorContext, farm_id: Option<&str>) -> bool {
    match farm_id {
        None => false, // missing farm scope → DENY (fail closed)
        Some(f) => ctx.owned_farm_ids.iter().any(|x| x == f) || ctx.memberships.iter().any(|(x, _)| x == f),
    }
}

fn member_with_role(ctx: &ActorContext, farm_id: Option<&str>, roles: &[&str]) -> bool {
    match farm_id {
        None => false,
        Some(f) => ctx.memberships.iter().any(|(x, r)| x == f && roles.contains(&r.as_str())),
    }
}

/// THE decision function. Unknown actions cannot be expressed in this enum,
/// which structurally guarantees fail-closed behaviour (mirrors KNOWN_ACTIONS).
pub fn can(ctx: &ActorContext, action: Action, farm_id: Option<&str>) -> bool {
    let admin = ctx.personas.iter().any(|p| p == "admin");
    if std::env::var("AUTHZ_DEBUG").is_ok() {
        println!("[authz-debug] action={:?} farm={:?} personas={:?} owned={:?} members={:?}", action, farm_id, ctx.personas, ctx.owned_farm_ids, ctx.memberships);
    }
    match action {
        Action::IssueView => admin || belongs_to_farm(ctx, farm_id),
        Action::IssueCreate | Action::IssueAdvance => {
            admin
                || member_with_role(ctx, farm_id, &["worker", "moderator"])
                || ctx.owned_farm_ids.iter().any(|x| Some(x.as_str()) == farm_id)
        }
        Action::IssueClose => {
            admin
                || member_with_role(ctx, farm_id, &["moderator"])
                || ctx.owned_farm_ids.iter().any(|x| Some(x.as_str()) == farm_id)
        }
        // ADR-017: physical actuation = moderator+ only, NEVER workers.
        Action::ValveControl => {
            admin
                || member_with_role(ctx, farm_id, &["moderator"])
                || ctx.owned_farm_ids.iter().any(|x| Some(x.as_str()) == farm_id)
        }
        // Authenticated read; per-item tenancy enforced handler-side via has_farm_access.
        Action::DeviceView => !ctx.personas.is_empty(),
        // Platform administration: admin-only by definition.
        Action::PlanManage | Action::SubscriptionAssign | Action::PersonaVerify
        | Action::AuditView | Action::FlagManage => admin,
    }
}

/// Handler-side tenancy assertion mirroring hasFarmAccess() in Node:
/// item-level reads check THIS instead of trusting list-level guards.
pub fn has_farm_access(ctx: &ActorContext, farm_id: Option<&str>) -> bool {
    belongs_to_farm(ctx, farm_id)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::store;

    #[test]
    fn worker_can_advance_issue_on_own_farm() {
        let db = store::seed();
        let ctx = build_actor_context(&db, "u-worker").unwrap();
        println!("personas={:?} members={:?} owned={:?}", ctx.personas, ctx.memberships, ctx.owned_farm_ids);
        assert!(can(&ctx, Action::IssueAdvance, Some("f-1")), "worker must advance issues on own farm");
    }
}
